ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx

```diff
@@ -48,94 +54,99 @@ function nodeLabel(n: NodeStatus) {
     };
 }
 
-const classNames = (() => {
-    const v: {[label: string]: boolean} = {
-        Artifact: true,
-        Suspended: true,
-        Collapsed: true
-    };
-    Object.entries(NODE_PHASE).forEach(([, label]) => (v[label] = true));
-    return v;
-})();
-
-export class WorkflowDag extends React.Component<WorkflowDagProps, WorkflowDagRenderOptions> {
-    private graph: Graph;
-
-    constructor(props: Readonly<WorkflowDagProps>) {
-        super(props);
-        this.state = {
-            expandNodes: new Set(),
-            showArtifacts: localStorage.getItem('showArtifacts') !== 'false'
-        };
-    }
+const classNames = {
+    Artifact: true,
+    Suspended: true,
+    Collapsed: true,
+    ...Object.entries(NODE_PHASE).reduce((acc, [, value]) => ({...acc, [value]: true}), {})
+};
 
-    private get artifactRepository() {
-        return this.props.artifactRepositoryRef?.artifactRepository || {};
-    }
+export function WorkflowDag(props: WorkflowDagProps) {
+    let graph: Graph;
+    const artifactRepository = props.artifactRepositoryRef?.artifactRepository || {};
+    const tags: Record<string, boolean> = Object.values(props.nodes || {}).reduce((acc, node) => ({...acc, [getNodeLabelTemplateName(node)]: true}), {});
 
-    public render() {
-        this.prepareGraph();
-
-        const tags: {[key: string]: boolean} = {};
-        Object.values(this.props.nodes || {}).forEach(n => (tags[getNodeLabelTemplateName(n)] = true));
-
-        return (
-            <GraphPanel
-                storageScope='workflow-dag'
-                graph={this.graph}
-                nodeGenresTitle={'Node Type'}
-                nodeGenres={genres}
-                nodeClassNamesTitle={'Node Phase'}
-                nodeClassNames={classNames}
-                nodeTagsTitle={'Template'}
-                nodeTags={tags}
-                nodeSize={this.props.nodeSize || 32}
-                defaultIconShape='circle'
-                hideNodeTypes={true}
-                hideOptions={this.props.hideOptions}
-                selectedNode={this.props.selectedNodeId}
-                onNodeSelect={id => this.selectNode(id)}
-                options={<WorkflowDagRenderOptionsPanel {...this.state} onChange={workflowDagRenderOptions => this.saveOptions(workflowDagRenderOptions)} />}
-            />
-        );
-    }
+    const [expandNodes, setExpandNodes] = useState(new Set(''));
```

Comment:
was there a reason you gave it the default `''`, different from the previous?

Reply:
In previous version,`expandNodes` would be of type unknown.
I wanted to gave it a more specific type, so assigned a default value.

Reply:
I decided to use generics here!